_posts/2019/2019-10-10-building-a-drawer-component-in-react.md

```diff
@@ -0,0 +1,805 @@
+
+  this.state = {
+    active: false,
+  };
+
+  this.handleTitleClick = this.handleTitleClick.bind(this);
+}
+```
+<blockquote class="Info Info-right">
+  "If you are using the experimental public class fields syntax, you can use class fields to correctly bind callbacks"
+  <cite>- <a href="https://reactjs.org/docs/handling-events.html">reactjs.org</a></cite>
+</blockquote>
+
+While this currently works, we can clean this up by taking it one step further using the **public class fields syntax**.
+
+I really love how easy this syntax reads and how it cleans up your constructor function from all the binding cruft.
+
+``` jsx
+constructor(props) {
+  super(props);
+
+  this.state = {
+    active: false,
+  };
+
+  // No binding necessary
+}
+
+// Instead of () {} is now = () => {}
+handleTitleClick = () => {
+  this.setState({ active: !this.state.active });
+}
+```
+
+So with all that out of the way, we now have our first iteration of our component below.
+
+``` jsx
+import React from 'react';
+
+class Drawer extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      active: false,
+    };
+  }
+
+  handleTitleClick = () => {
+    this.setState(prevState => ({ active: !prevState.active }));
+  };
+
+  render() {
+    const { title, children } = this.props;
+    const drawerStyles = this.state.active ? 'is-expanded' : '';
+
+    return (
+      <div className={`Drawer ${drawerStyles}`}>
+        <div
+          className="Drawer-title"
+          onClick={this.handleTitleClick}
+        >
+          {title}
+        </div>
+        <div className="Drawer-content">
+          {children}
+        </div>
+      </div>
+    );
+  }
+}
+
+export default Drawer;
+
+/////////////////////////////////////
+// And here's how you would use it //
+/////////////////////////////////////
+import React from 'react';
+import Drawer from '~/components/Drawer'; // The location of our new component
+
+class AnotherComponent extends React.Component {
+  // code omitted
+  render() {
+    return (
+      <div className="AnotherComponent">
+        <h1>Welcome to my component</h1>
+        <Drawer title="My Drawer Title">
+          <ul>
+            <li>Item 1</li>
+            <li>Item 2</li>
+            <li>Item 3</li>
+          </ul>
+        </Drawer>
+      </div>
+    );
+  }
+}
+```
+
+Here's what it looks like so far. Really the only thing it does is provide
+structure and toggle a CSS class `is-expanded` based on state.
+
+![Unstyled Drawer Component](/img/2019/unstyled-drawer-component.gif)
+
+It's pretty ugly at this point, so in the next section we're going to add some
+SCSS styles to it.
+
+## Adding Style
+
+One thing missing from our component at the moment is that the Drawer's content
+is always shown. We only want it to display if the state of the component is expanded. We're going to eventually animate it into existing using a slide transition so focusing on the element's height will be advantageous here.
+
```

Comment:
One problem right now, is the drawer is basically always open. This means the Drawer's content
is always shown. We only want it to display if the state of the component is expanded. Just like a real drawer, we want to be able to open and close it or show and hide the content as needed. 


\/ this feels out of left field. is this the next thing you're working on? or something you'll do later? 
We're going to eventually animate it into existing using a slide transition so focusing on the element's height will be advantageous here.